File: sem4/jpp/l3/rust/src/dhsetup.rs
use rand::Rng;
use crate::gfint::{self, GFInt};

#[derive(Clone)]
pub struct DHSetup<const N: i32> {
    generator: gfint::GFInt<N>
}

impl<const N: i32> DHSetup<N> {

    fn power_mod(mut x: i64, mut y: i64, p: i64) -> i32 {
        let mut res = 1;
        x = x % p;
        while y > 0 {
            if y & 1 == 1 {
                res = (res * x) % p;
            }
            y = y >> 1;
            x = ( x * x) % p;
        }
        res as i32
    }

    fn prime_factors(mut n: i32) -> Vec<i32> {
        let mut factors: Vec<i32> = vec![];
        while n % 2 == 0 {
            factors.push(2);
            n /= 2;
        }
        let sqrt = (n as f64).sqrt() as i32;
        for i in (3..=sqrt).step_by(2) {
            while n % i32::from(i) == 0 {
                factors.push(i32::from(i));
                n /= i32::from(i);
            }
        }
        if n > 2 {
            factors.push(n);
        }
        factors
    }

    fn generate_generator() -> GFInt<N> {
        let generator: GFInt<N>;
        let mut rng = rand::thread_rng();
        let factors = Self::prime_factors(N - 1);
        loop {
            let candidate = rng.gen_range(1..=N - 1);
            let mut found = true;
            for factor in &factors {
                if Self::power_mod(candidate.into(), ((N - 1) / *factor).into(), N.into()) == 1 {
                    found = false;
                    break;
                }
            }
            if found {
                generator = gfint::GFInt::new(candidate);
                return generator;
            }
        }
    }

    pub fn new() -> Self {
        DHSetup {
            generator: Self::generate_generator()
        }
    }

    pub fn get_generator(self) -> GFInt<N> {
        self.generator
    }

    pub fn power(self, a: GFInt<N>, b: u128 ) -> GFInt<N> {
        let mut result: GFInt<N> = GFInt::new(1);
        for _ in 0..b {
            result = result * a;
        }
        result
    } 
}

File: sem4/jpp/l3/java/DHSetup.java
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class DHSetup {
    private long CHARACTERISTIC;
    private GF_Int generator;
    private Random random = new Random();


    public long getCharacteristic() {
        return CHARACTERISTIC;
    }

    public DHSetup(long characteristic) {
        this.CHARACTERISTIC = characteristic;
        generateGenerator();
    }

    private long powerMod(long x, long y, long p) {
        if (y == 0) {
            return x;
        }

        if (x == 0) {
            return y;
        }
        long res = 1;
        x = x % p;
        if (x < 0) x += p;
        while (y > 0) {
            if ((y & 1) == 1) {
                res = (res * x) % p;
            }
            y = y >> 1;
            x = (x * x) % p;
            if (x < 0) x += p;
        }
        return res;
    }

    public boolean isGenerator(long g, long p) {
        List<Long> factors = primeFactors(p - 1);
        for (long factor : factors) {
            if (powerMod(g, (p - 1) / factor, p) == 1) {
                return false;
            }
        }
        return true;
    }

    private List<Long> primeFactors(long n) {
        List<Long> factors = new ArrayList<>();
        while (n % 2 == 0) {
            factors.add(2L);
            n = n / 2;
        }
        for (long i = 3; i <= Math.sqrt(n); i = i + 2) {
            while (n % i == 0) {
                factors.add(i);
                n = n / i;
            }
        }
        if (n > 2)
            factors.add(n);
        return factors;
    }

    private void generateGenerator() {
        List<Long> factors = primeFactors(CHARACTERISTIC - 1);

        while (true) {
            long candidate = 2L + (long) (Math.random() * (CHARACTERISTIC - 2));
            boolean found = true;
            for (long factor : factors) {
                if (powerMod(candidate, (CHARACTERISTIC - 1) / factor, CHARACTERISTIC) == 1) {
                    found = false;
                    break;
                }
            }
            if (found) {
                generator = new GF_Int(candidate, CHARACTERISTIC);
                return;
            }
        }
    }

    public GF_Int power(GF_Int a, long b) {
        GF_Int result = new GF_Int(1, CHARACTERISTIC);
        GF_Int base = a;
    
        while (b > 0) {
            if (b % 2 != 0) {
                result = result.multiply(base);
                result = result.mod();
            }
            base = base.multiply(base);
            b /= 2;
        }
    
        return result;
    }

    public GF_Int getGenerator() {
        return generator;
    }

}
